typescript-typing-tutor/main.ts

```diff
@@ -0,0 +1,101 @@
+const typeSentences = [
+  'wizards use intelligence and sorcerers use charisma',
+  'drow elves have superior darkvision',
+  'kenku have excellent memory',
+  'werewolves are chaotic evil',
+  'homunculi are loyal servants',
+  'halflings are very lucky creatures',
+];
+
+function getNewSentence(): void {
+  if (!$textBox) throw new Error('$textBox query failed');
+
+  const randInt = Math.floor(Math.random() * typeSentences.length);
+  currentSentence = typeSentences[randInt];
+
+  while ($textBox.firstChild) {
+    $textBox.removeChild($textBox.lastChild as Node);
+  }
+
+  const $nullSpan = document.createElement('span');
+  $nullSpan.id = 'null-span';
+  $textBox.appendChild($nullSpan);
+
+  for (let i = 0; i < currentSentence.length; i++) {
+    const $letterSpan = document.createElement('span');
+    $letterSpan.textContent = currentSentence[i];
+    $letterSpan.id = 'letter' + String(i);
+    $letterSpan.className = 'untyped';
+    $textBox.appendChild($letterSpan);
+  }
+}
+
+function updateCurrentSpan(): void {
+  $currentSpan = document.querySelector(`#letter${currentLetterNum}`);
+  if (!$currentSpan)
+    throw new Error('$currentSpan query failed at updateCurrentSpan');
+  $currentSpan.className = 'current untyped';
+}
+
+let currentSentence: string;
+let currentLetterNum = 0;
+
+const $textBox = document.querySelector('.textbox');
+const $restartButton = document.querySelector('#restart');
+const $accuracy = document.querySelector('#accuracy');
+let $currentSpan = document.querySelector(`#letter${currentLetterNum}`);
+
+if (!$textBox) throw new Error('$textBox query failed');
+if (!$restartButton) throw new Error('$restartButton query failed');
+if (!$accuracy) throw new Error('$restartButton query failed');
+
+getNewSentence();
+updateCurrentSpan();
+
+let totalKeyPress = 0;
+let correctKeyPress = 0;
+
+document.addEventListener('keydown', (event: KeyboardEvent) => {
+  if (!$currentSpan)
+    throw new Error('$currentSpan query failed at event start');
+
+  const focusLetter = $currentSpan.textContent;
+
+  if (event.key === focusLetter) {
+    $currentSpan.className = 'correct';
+
+    currentLetterNum++;
+
+    if (currentLetterNum < currentSentence.length) {
+      updateCurrentSpan();
+      correctKeyPress++;
+    } else {
+      $currentSpan = document.querySelector('#null-span');
+    }
+  } else {
+    $currentSpan.className = 'current incorrect';
+  }
```

Comment:
I think we could make this section a littler cleared. Since we know that if the input is _incorrect_, we would want to set it to incorrect and stop the rest of the function because we do not continue unless the user is correct. If we were to use a guard clause for incorrect, we could then return out of the function after setting the `className` to incorrect and pull out code out of the `if` block and then no long have a nested `if` statement. 

For example:
```
if (event.key !== focusLetter) {
  $currentSpan.className = 'current incorrect';
  return;
}

$currentSpan.className = 'correct';
currentLetterNum++;

if (currentLetterNum < currentSentence.length) { ...
```